templates/_helpers.tpl

```diff
@@ -0,0 +1,242 @@
+  {{- end -}}
+{{- end -}}
+
+{{/*
+Generate olcSyncRepl list
+*/}}
+{{- define "olcSyncRepls2" -}}
+{{- $name := (include "openldap.fullname" .) }}
+{{- $domain := (include "global.baseDomain" .) }}
+{{- $bindDNUser := .Values.global.adminUser }}
+{{- $namespace := .Release.Namespace }}
+{{- $cluster := .Values.replication.clusterName }}
+{{- $adminPassword := ternary .Values.global.adminPassword "%%ADMIN_PASSWORD%%" (empty .Values.global.existingSecret) }}
+{{- $retry := .Values.replication.retry }}
+{{- $timeout := .Values.replication.timeout }}
+{{- $starttls := .Values.replication.starttls }}
+{{- $tls_reqcert := .Values.replication.tls_reqcert }}
+{{- $interval := .Values.replication.interval }}
+{{- $nodeCount := .Values.replicaCount | int }}
+{{- $port := .Values.global.ldapPort | int }}
+  {{- range $index0 := until $nodeCount }}
+    {{- $index1 := $index0 | add1 }}
+    olcSyncrepl:
+      rid=10{{ $index1 }}
+      provider=ldap://{{ $name }}-{{ $index0 }}.{{ $name }}-headless.{{$namespace }}.svc.{{ $cluster }}:{{ $port }}
+      binddn={{ printf "cn=%s,%s" $bindDNUser $domain }}
+      bindmethod=simple
+      credentials={{ $adminPassword }}
+      searchbase={{ $domain }}
+      type=refreshAndPersist
+      interval={{ $interval }}
+      network-timeout=0
+      retry="{{ $retry }} +"
+      timeout={{ $timeout }}
+      starttls={{ $starttls }}
+      tls_reqcert={{ $tls_reqcert }}
+  {{- end -}}
+{{- end -}}
+
+{{/*
+Renders a value that contains template.
+Usage:
+{{ include "openldap.tplValue" ( dict "value" .Values.path.to.the.Value "context" $) }}
+*/}}
+{{- define "openldap.tplValue" -}}
+    {{- if typeIs "string" .value }}
+	{{- tpl .value .context }}
+    {{- else }}
+	{{- tpl (.value | toYaml) .context }}
+    {{- end }}
+{{- end -}}
+
+{{/*
+Return the proper Openldap image name
+*/}}
+{{- define "openldap.image" -}}
+{{- include "common.images.image" (dict "imageRoot" .Values.image "global" .Values.global) -}}
+{{- end -}}
+
+{{/*
+Return the proper Docker Image Registry Secret Names
+*/}}
+{{- define "openldap.imagePullSecrets" -}}
+{{ include "common.images.pullSecrets" (dict "images" (list .Values.image ) "global" .Values.global) }}
+{{- end -}}
+
+
+{{/*
+Return the proper Openldap init container image name
+*/}}
+{{- define "openldap.initContainerImage" -}}
+{{- include "common.images.image" (dict "imageRoot" .Values.customTLS.image "global" .Values.global) -}}
+{{- end -}}
+
+
+{{/*
+Return the list of builtin schema files to mount
+Cannot return list => return string comma separated
+*/}}
+{{- define "openldap.builtinSchemaFiles" -}}
+  {{- $schemas := "" -}}
+  {{- if .Values.replication.enabled -}}
+    {{- $schemas = "syncprov,serverid,csyncprov,rep,bsyncprov,brep,acls" -}}
```

Comment:
Do you need to pretend all of these (config changes) are schemas to smuggle them in? If so, you should make changes to the container so there's a proper way of doing this kind of thing and use it here.

Reply:
There are three ways before `slapd` starts and listens on TCP port(s) making it "available for service" in the eyes of Kubernetes and one after:
* extra config
* custom schema
* custom schemas
and after
* ldap custom init scripts

What's better here?  I'd imagine that having these loaded into the server before it opens the TCP port, so traffic doesn't start arriving when replication isn't setup but you tell me.

Reply:
Depends, do you want to rely on the API that the container provides? Then extend that in ways that make sense (at least) for your usecase. Also if the server appears live in Kubernetes before "custom init scripts" are even run, then that sounds like a problem anyway.

Reply:
fixed, they happen much later in a more obvious location now